snewpdag/plugins/Validator.py

```diff
@@ -0,0 +1,75 @@
+'''
+This will check that a field has the proper form,
+and either clean it up or consume the action.
+Then we don't have to have each downstream plugin
+repeat this sort of validation.
+
+check_key() checks that the desired key/field exists in the payload
+
+check_type() checks that the field has the correct type,
+e.g. 'action': str, 'history': History
+
+check_listtype() checks the type of the elements in the list
+based on the type we want in the input,
+and if a wrong type is found, such element is removed
+and the removal is logged.
+
+check_sorted() is for numbers in a list.
+It checks whether it is sorted in ascending/descending order.
+If it is not sorted, then it sorts the list based on the order in the input.
+We may also pass the list unmodified by not including order in the input.
+
+'''
+
+import logging
+import numpy as np
+import healpy as hp
+from snewpdag.dag import Node
+
+class Validator(Node):
+    def __init__(self, **kwargs):
+        self.search_key = kwargs.pop('search_key', None)
+        self.key_type = kwargs.pop('key_type', None)
+        self.order = kwargs.pop('order', None)
+        super().__init__(**kwargs)
+    
+    def check_key(self, data): # check that the key exists
+        if self.search_key in data:
+            return True
+        else:
+            logging.error('Desired key [{}] not found in payload'.format(self.search_key))
+            return False
+
+    def check_type(self, data): # check that the key corresponed to the correct type
+        return isinstance(data, self.key_type)
```

Comment:
Does this work? self.key_type will be a string, but I think arg2 needs to be the type/class

Reply:
All the unit tests seem to be working with no errors when I use self.keytype for isinstance(), so I suppose it does work!

Reply:
I think my unit tests work because I didn't treat self.key_type itself as a string, but rather as a type per se.

For instance, when calling the plugin in the **unit test**, instead of writing
     Validator_keytype( key_type = 'str' )         # where the class of key_type itself is a string here,
I wrote 
     Validator_keytype( key_type = str )          # where the class of key_type itself is a type here.

However, I'm not sure if this is legal, but I suppose I can avoid this problem in **the plugin itself** by checking the type conditions one-by-one instead of using an input, e.g.
     if data = 'action'
          return isinstance(data, str)
     elif data = 'dt'
          return isintance(data, float)
     elif
     .
     .
     .

Though I imagine this would be a bit more tedious (I'm not too sure if this is what you meant in the meeting today!)